### ISSUE TITLE
BKSF mapper fails to map H2 operator from driver

### ISSUE THREAD

**Issue by woodsp-ibm** [Qiskit/qiskit-nature#436]
### Environment

- **Qiskit Nature version**: `main` branch (did not try release)
- **Python version**: 3.6
- **Operating system**: Linux

### What is happening?

In #432 I was asked had I tried with BKSF as I had added Parity and BK mappers into the [test_excited_states_solvers.py](https://github.com/Qiskit/qiskit-nature/blob/main/test/algorithms/excited_state_solvers/test_excited_states_solvers.py) test case when fixing the referenced issue and observing issues around two qubit reduction and symmetries. I did not try BKSF, you can see my comment in the referenced PR, but in trying it fails in the ground state computation. Now the bksf test file includes an pre-defined operator that is commented as H2 - in this case its built from the driver and here it fails with the following exception.

```
<snip>
line 480, in _add_edges_for_term
    _type = _interaction_type(n_number, n_raise, n_lower)
  File "...qiskit_nature/mappers/second_quantization/bksf.py", line 421, in _interaction_type
    raise ValueError("unexpected number of number operators: ", n_number)
ValueError: ('unexpected number of number operators: ', 0)
```

### How can we reproduce the issue?

Based on the 
```
from qiskit import BasicAer
from qiskit.utils import algorithm_globals, QuantumInstance
from qiskit_nature.drivers import UnitsType
from qiskit_nature.drivers.second_quantization import PySCFDriver
from qiskit_nature.mappers.second_quantization import BravyiKitaevSuperFastMapper, BravyiKitaevMapper
from qiskit_nature.converters.second_quantization import QubitConverter
from qiskit_nature.problems.second_quantization import ElectronicStructureProblem
from qiskit_nature.algorithms import  GroundStateEigensolver, VQEUCCFactory
 


algorithm_globals.random_seed = 8
driver = PySCFDriver(
    atom="H .0 .0 .0; H .0 .0 0.75",
    unit=UnitsType.ANGSTROM,
    charge=0,
    spin=0,
    basis="sto3g",
)
es_problem = ElectronicStructureProblem(driver)
quantum_instance = QuantumInstance(
    BasicAer.get_backend("statevector_simulator"),
    seed_transpiler=90,
    seed_simulator=12,
)


#converter = QubitConverter(BravyiKitaevMapper())
converter = QubitConverter(BravyiKitaevSuperFastMapper())
solver = VQEUCCFactory(quantum_instance)
gsc = GroundStateEigensolver(converter, solver)

result = gsc.solve(es_problem)
print(result)
```

### What should happen?

If the code above is switched to say the BravyiKitaevMapper (not the super fast one) then it works and prints the result as below. I would expect the same with BKSF
```
=== GROUND STATE ENERGY ===
 
* Electronic ground state energy (Hartree): -1.842686681847
  - computed part:      -1.842686681847
~ Nuclear repulsion energy (Hartree): 0.70556961456
> Total ground state energy (Hartree): -1.137117067287
 
=== MEASURED OBSERVABLES ===
 
  0:  # Particles: 2.000 S: 0.000 S^2: 0.000 M: 0.000
 
=== DIPOLE MOMENTS ===
 
~ Nuclear dipole moment (a.u.): [0.0  0.0  1.41729459]
 
  0: 
  * Electronic dipole moment (a.u.): [0.0  0.0  1.41729468]
    - computed part:      [0.0  0.0  1.41729468]
  > Dipole moment (a.u.): [0.0  0.0  -0.00000009]  Total: 0.00000009
                 (debye): [0.0  0.0  -0.00000024]  Total: 0.00000024
```

**Comment by mrossinek** [comment]
This appears to boil down to the fact that the `BKSFMapper` is unable to handle all-identity operators like `IIII`.
@jlapeyre could you take a look at this one?

**Comment by jlapeyre** [comment]
In the case given in the example above the problem can be solved by checking that coefficient is zero. Then we can skip analyzing the term.  A check for zero coefficient should be added just to save some computation.

But, I'll have to do something more to solve the general problem.

**Comment by jlapeyre** [comment]
I've made a few changes needed for corner cases.
But, running the example above, I encounter a Fermionic Hamiltonian
that looks like this
```
(1+0j) * (+_0 +_2)
```
which is converted correctly to the dense operator
```
(1+0j) * (+I+I)
```

I don't think the the electronic structure problems should produce terms with unbalanced raising and lowering operators. In fact this is not just a term, but the entire Hamiltonian, and it's not Hermitian.

I am surprised that `BravyiKitaevMapper()` does not error. Does it give a sensible result?

I am working mostly from [this paper](https://arxiv.org/abs/1712.00446), which can't handle a term like the one above.

**Comment by jlapeyre** [comment]
It looks like we always handle the "nuclear repulsion" energy only at the very end of the VQE. This makes sense. But, we should probably avoid constructing input Fermionic operators that have a term proportional to the identity, even if the coefficient is zero. I don't know if this is a bug or an oversight.

**Comment by woodsp-ibm** [comment]
> I am surprised that BravyiKitaevMapper() does not error. Does it give a sensible result?

The sample code above, that would reproduce the reported failure, shows the result as output from when using BK mapper. I think this is consistent with Parity or JW mappings i.e. sensible

**Comment by jlapeyre** [comment]
Ok. I'll compare with the other mappings.

**Comment by mrossinek** [comment]
> But, running the example above, I encounter a Fermionic Hamiltonian that looks like this
> 
> ```
> (1+0j) * (+_0 +_2)
> ```
> 
> which is converted correctly to the dense operator
> 
> ```
> (1+0j) * (+I+I)
> ```
> 
> I don't think the the electronic structure problems should produce terms with unbalanced raising and lowering operators. In fact this is not just a term, but the entire Hamiltonian, and it's not Hermitian.

Hi @jlapeyre, thanks for looking into this! What you're observing here is actually the `FermionicOp`-representation of an initial `HartreeFock` state. You're right, that this is not a Hamiltonian which we expect to encounter.
For the HF initial state we need to initialize the occupied qubits according to the occupied spin orbitals in the second quantization space. Since refactoring to our more modular mapping approach, we are leveraging the `mappers` module to compute which qubits these locations correspond to here:

https://github.com/Qiskit/qiskit-nature/blob/9b7e2abf0a5190128e7b0be276c020195e244a00/qiskit_nature/circuit/library/initial_states/hartree_fock.py#L81-L99

I am not entirely sure how we used to handle this in the past in Aqua when using the `BKSF` mapper. It actually looks like we may have not supported the `bksf` case in Aqua at all:

https://github.com/Qiskit/qiskit-aqua/blob/7c2c2e546e9002e758ff6b25f98a9a594589bd6f/qiskit/chemistry/circuit/library/initial_states/hartree_fock.py#L106-L126

@pbark @woodsp-ibm please correct me if I am wrong. Also, what do you suggest in terms of how to proceed here?

**Comment by woodsp-ibm** [comment]
In Aqua we had the JW, Parity and BK mappers, which the UCCSD, HF state and core supported via strings denoting which mapping was used and the HF state was computed directly from knowing the mapping denoted by the string - as you have indicated by your link above. The BKSF mapper was added later, after the aforementioned code was already there. It could be used to map a FermionicOperator but was never integrated to the same extent - this mapping is based on interactions and quickly led to large numbers of qubits, so seemed less practical in the short term for sort of problems the stack was addressing, so it was left for more 'manual' use within the stack where you could create a FermionicOperator from a driver and map it to a qubit operator, so you could use it to experiment with but its integration was not done to the same extent - priorities, other work etc

**Comment by mrossinek** [comment]
Thanks, @woodsp-ibm, for the insight! I also had a quick discussion with @pbark offline and we think, that for the time being, we should simply fix as much as possible (as done in #451) and add a check inside of `HartreeFock` which raises a useful error to the user when detecting the `BKSF` mapper, preventing the more cryptic error message which we currently observe.

@jlapeyre is that something you can add as part of #451?

**Comment by jlapeyre** [comment]
> @jlapeyre is that something you can add as part of #451?

Yes, I pushed a commit to #451 doing this.
